File: handlers/status.ts
import {
  Composer,
  InputFile,
  InlineKeyboard,
} from "https://deno.land/x/grammy/mod.ts";
import { getHistory } from "../api.ts";
import { getAccessToken } from "../db.ts";
import { generate } from "../image.ts";
import { getReplyMarkup } from "../utils.ts";
import env from "../env.ts";

const composer = new Composer();

composer.hears(
  new RegExp(`^/(status|status@${env.BOT_USERNAME})$`, "i"),
  async (ctx) => {
    if (!ctx.from || !ctx.message.text) {
      return;
    }

    const access = await getAccessToken(ctx.from.id);

    if (!access) {
      await ctx.reply(
        "You need to connect your Deezer account first. PM me and use the /connect command.",
        {
          reply_markup: new InlineKeyboard().url(
            "PM me",
            `https://t.me/${ctx.me.username}`
          ),
        }
      );
      return;
    }

    let indent = Number(ctx.message.text.split(/\s/)[1]);
    const history = await getHistory(access);

    if (!(indent - 1 in history)) indent = 0;
    else indent -= 1;

    const track = history[indent];

    await ctx.replyWithPhoto(
      new InputFile(
        await generate(
          track.album.cover_big,
          ctx.from.first_name,
          track.title,
          track.artist.name,
          track.album.title
        )
      ),
      { reply_markup: getReplyMarkup(track) }
    );
  }
);

export default composer;

File: handlers/status.go
package handlers

import (
	"errors"
	"fmt"
	"strconv"
	"strings"

	"github.com/roj1512/DeezerHistory/api"
	"github.com/roj1512/DeezerHistory/helpers"
	"github.com/roj1512/DeezerHistory/photo"

	"github.com/PaulSonOfLars/gotgbot/v2"
	"github.com/PaulSonOfLars/gotgbot/v2/ext"
	"github.com/gotgbot/keyboard"
)

func statusHandler(b *gotgbot.Bot, ctx *ext.Context) error {
	tracks, err := api.FetchHistory(ctx.EffectiveUser.Id)
	if err != nil {
		if errors.Is(err, api.ErrNotAuthorized) {
			_, err := ctx.EffectiveMessage.Reply(b, "You’re not authorized!", &gotgbot.SendMessageOpts{
				ReplyMarkup: new(
					keyboard.InlineKeyboard,
				).Url(
					"Authorize",
					fmt.Sprintf("https://t.me/%s?start=connect", b.Username),
				).Build(),
			})
			return err
		}
		return err
	}
	if len(tracks) == 0 {
		_, err = ctx.EditedMessage.Reply(b, "Your history is empty!", nil)
		return err
	}
	indent := 0
	args := strings.Fields(ctx.EffectiveMessage.Text)
	if len(args) > 1 {
		indent, _ = strconv.Atoi(args[1])
	}
	indent = helpers.EnsureIndent(indent, len(tracks))
	track := tracks[indent]
	photo, err := photo.Generate(track.Album.Title, track.Album.CoverBig, track.Artist.Name, track.Title, ctx.EffectiveUser.FirstName)
	if err != nil {
		return err
	}
	_, err = b.SendPhoto(ctx.EffectiveChat.Id, photo, &gotgbot.SendPhotoOpts{
		ReplyMarkup: helpers.GetReplyMarkup(track),
	})
	return err
}
